rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,206 @@
+
+The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component
+
+```typescript
+export const Sample = React.forwardRef<HTMLElement, SampleProps>((props, ref) => {
+  const state = useSample(props, ref);
+  useSampleStyles(state);
+
+  return renderSample(state);
+});
+```
+
+#### useSample.ts
+
+Accepts the component props and handles the internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props, we set defaults internally but if consumer wants component variants with different defaults, they can set them
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps): SampleState => {
+  const resolvedRef = ref || React.useRef();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+This should be a pure function whose sole responsibility is to render JSX from the provided state. No state mutation or processing
+should happen in this function, but rather done in the `useSample` hook
+
+```typescript
+import { getSlots } from '@fluentui/react-utils'
+
+// state should come from `useSample` hook
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
```

Comment:
```suggestion
import { getSlots } from '@fluentui/react-utils';
import { sampleShorthandProps } from './useSample';

// state should come from `useSample` hook
export const renderSample = (state: SampleState) => {
  const { slots, slotProps } = getSlots(state, sampleShorthandProps);
```

Reply:
Since we defined `sampleShorthandProps` to be nonempty in `useSample.ts`, we should include that in here. Also, I think it might be worth pointing out that when the component is not supposed to have any slots, we should omit shorthand props altogether like in this case, and have simply `const { slots, slotProps } = getSlots(state);`.

Reply:
Makes sense

Reply:
Let's default with shorthand, my intention is to make the non-slots scenario obvious with the way the code samples are presented... or at least easy to figure that out